File: CPPVersion/AoC2020CPP/AoC2020CPP/Day13.cpp
#include "Day13.h"

Day13::Day13() : DefaultClass(13, true)
{
	std::vector<std::string> inp = FR.ReadLine();
	BusInp = FR.SplitString(inp[1], ',');
	for (int i = 0; i < BusInp.size(); i++)
	{
		if (BusInp[i] != "x")
			BusID.push_back(std::stoi(BusInp[i]));
	}
	ResetClock();
	Part1(std::stoi(inp[0]));
	TimeResetClock();
	Part2();
	TimeClock();

}

void Day13::Part1(int Arrival)
{
	int ID = 0, Wait = INT16_MAX;
	for (int i = 0; i < BusID.size(); i++)
	{
		int temp = BusID[i] * (Arrival / BusID[i] + 1) - Arrival;
		if (temp < Wait)
		{
			Wait = temp;
			ID = BusID[i];
		}
	}
	std::cout << "Part1: " << Wait * ID<<"\n";
}

void Day13::Part2()
{
	int minute = 0;
	std::vector<int> CRH = std::vector<int>();
	int x = 0;
	for (int i = 0; i < BusInp.size(); i++)
	{
		if (BusInp[i] != "x")
		{
			CRH.push_back(BusID[x] - (minute % BusID[x]));
			x++;
		}
		minute++;
	}
	long long timestamp = 0;
	long long multiplier = BusID[0];
	for (int i = 1; i < BusID.size(); i++)
	{	
		while ((timestamp % BusID[i]) != CRH[i])
			timestamp += multiplier;
		multiplier *= BusID[i];
	}
	std::cout << "Part2: " << timestamp << "\n";
}

File: JavaVersion/AoC2020/src/com/company/Day13.java
package com.company;

import java.util.ArrayList;

public class Day13 extends DefaultClass{

    public Day13()
    {
        super(13,true);
        ArrayList<String> lines = fr.ReadLine();
        int arrival = Integer.parseInt(lines.get(0));
        ArrayList<String> ID = fr.SplitString(lines.get(1),",");
        int BusID=0, wait=Integer.MAX_VALUE;
        for (int i = 0; i < ID.size(); i++) {
            if(!ID.get(i).equals("x"))
            {
                int div = arrival/Integer.parseInt(ID.get(i));
                int temp = Integer.parseInt(ID.get(i))  *(div+1)-arrival;
                if(temp<wait)
                {
                    wait = temp;
                    BusID = Integer.parseInt(ID.get(i));
                }
            }
        }
        System.out.println("Part1: " + (wait*BusID));
        ArrayList<Integer> first = new ArrayList<>();
        ArrayList<Integer> seconds = new ArrayList<>();
        int minute = -1;
        for (int i = 0; i < ID.size(); i++) {
            minute++;
            if(!ID.get(i).equals("x"))
            {
                first.add(Integer.parseInt(ID.get(i))-(minute%Integer.parseInt(ID.get(i))));
                seconds.add(Integer.parseInt(ID.get(i)));
            }
        }
        long timestamp = 0;
        long multiplier = seconds.get(0);
        for (int i = 1; i < first.size(); i++) {
            while((timestamp%seconds.get(i))!= first.get(i))
            {
                timestamp+=multiplier;
            }
            multiplier*=seconds.get(i);
        }
        System.out.println("Part2: "+timestamp);

    }
}
